File: archivist/src/test/java/com/zorroa/archivist/service/UserServiceTests.java
package com.zorroa.archivist.service;

import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Lists;
import com.zorroa.archivist.AbstractTest;
import com.zorroa.archivist.domain.*;
import com.zorroa.sdk.client.exception.DuplicateElementException;
import org.junit.Before;
import org.junit.Test;

import java.util.List;

import static org.junit.Assert.*;

/**
 * Created by chambers on 12/23/15.
 */
public class UserServiceTests extends AbstractTest {

    User testUser;

    @Before
    public void init() {
        UserSpec builder = new UserSpec();
        builder.setUsername("billybob");
        builder.setPassword("123password!");
        builder.setEmail("testing@testing123.com");
        builder.setFirstName("BillyBob");
        builder.setLastName("Rodriquez");
        testUser = userService.create(builder);
    }

    @Test
    public void createUser() {
        UserSpec builder = new UserSpec();
        builder.setUsername("test");
        builder.setPassword("123password");
        builder.setEmail("test@test.com");
        builder.setFirstName("Bilbo");
        builder.setLastName("Baggings");
        builder.setPermissions(
                userService.getPermission("group::manager"));
        User user = userService.create(builder);

        /*
         * Re-authenticate this test as the user we just made.  Otherwise we can't
         * access the user folder.
         */
        authenticate(user.getUsername());

        /*
         * Get the user's personal folder.
         */
        Folder folder = folderService.get("/Users/test");
        assertEquals(folder.getName(), user.getUsername());

        /*
         * Get the permissions for the user.  This should contain the user's
         * personal permission and the user group
         */
        List<Permission> perms = userService.getPermissions(user);
        assertEquals(3, perms.size());
        assertTrue(userService.hasPermission(user, userService.getPermission("group::manager")));
        assertTrue(userService.hasPermission(user, userService.getPermission("user::test")));
        assertFalse(userService.hasPermission(user, userService.getPermission("group::developer")));
    }

    @Test(expected=DuplicateElementException.class)
    public void createDuplicateUser() {
        UserSpec builder = new UserSpec();
        builder.setUsername("test");
        builder.setPassword("123password");
        builder.setEmail("test@test.com");
        builder.setFirstName("Bilbo");
        builder.setLastName("Baggings");
        builder.setPermissions(
                userService.getPermission("group::manager"));
        userService.create(builder);
        userService.create(builder);
    }

    @Test
    public void createUserWithPresets() {
        UserPreset presets = userService.createUserPreset(new UserPresetSpec()
                .setName("defaults")
                .setPermissionIds(Lists.newArrayList(userService.getPermission("group::manager").getId()))
                .setSettings(new UserSettings().setSearch(ImmutableMap.of("foo", "bar"))));

        UserSpec builder = new UserSpec();
        builder.setUsername("bilbo");
        builder.setPassword("123password");
        builder.setEmail("bilbo@test.com");
        builder.setFirstName("Bilbo");
        builder.setLastName("Baggings");
        builder.setUserPresetId(presets.getPresetId());

        User user = userService.create(builder);
        assertTrue(userService.hasPermission(user, userService.getPermission("group::manager")));

        UserSettings settings = user.getSettings();
        assertEquals("bar", settings.getSearch().get("foo"));
    }

    @Test
    public void testSetEnabled() {
        UserSpec builder = new UserSpec();
        builder.setUsername("test");
        builder.setPassword("123password");
        builder.setEmail("test@test.com");
        builder.setFirstName("Bilbo");
        builder.setLastName("Baggings");
        builder.setPermissions(
                userService.getPermission("group::manager"));
        User user = userService.create(builder);

        assertTrue(userService.setEnabled(user, false));
    }

    @Test
    public void updatePermissions() {
        UserSpec builder = new UserSpec();
        builder.setUsername("bilbob");
        builder.setPassword("123password");
        builder.setEmail("test@test.com");
        builder.setFirstName("Bilbo");
        builder.setLastName("Baggings");
        builder.setPermissions(userService.getPermission("group::manager"));
        User user = userService.create(builder);

        Permission dev = userService.getPermission("group::developer");
        assertTrue(userService.hasPermission(user, userService.getPermission("group::manager")));
        assertFalse(userService.hasPermission(user, dev));

        userService.setPermissions(user, Lists.newArrayList(dev));

        assertFalse(userService.hasPermission(user, userService.getPermission("group::manager")));
        assertTrue(userService.hasPermission(user, dev));
    }

    @Test(expected = IllegalArgumentException.class)
    public void updatePasswordFailureTooShort() {
        try {
            userService.resetPassword(testUser, "nogood");
        } catch (IllegalArgumentException e) {
            assertTrue(e.getMessage().contains("or more characters"));
            throw e;
        }
    }

    @Test(expected = IllegalArgumentException.class)
    public void updatePasswordFailureNoCaps() {
        try {
            userService.resetPassword(testUser, "nogood");
        } catch (IllegalArgumentException e) {
            assertTrue(e.getMessage().contains("upper case"));
            throw e;
        }
    }

    @Test(expected = IllegalArgumentException.class)
    public void updatePasswordFailureNoNumbers() {
        try {
            userService.resetPassword(testUser, "nogood");
        } catch (IllegalArgumentException e) {
            assertTrue(e.getMessage().contains("least 1 number."));
            throw e;
        }
    }

    @Test
    public void updatePassword() {
        userService.resetPassword(testUser, "DogCatched1");
        userService.checkPassword(testUser.getUsername(), "DogCatched1");
    }

    @Test
    public void updateUser() {
        UserSpec builder = new UserSpec();
        builder.setUsername("bilbob");
        builder.setPassword("123password");
        builder.setEmail("test@test.com");
        builder.setFirstName("Bilbo");
        builder.setLastName("Baggings");
        User user = userService.create(builder);

        UserProfileUpdate update = new UserProfileUpdate();
        update.setFirstName("foo");
        update.setLastName("bar");
        update.setEmail("test@test.com");

        assertTrue(userService.update(user, update));
        User updated = userService.get(user.getId());
        assertEquals(update.getEmail(), updated.getEmail());
        assertEquals(update.getFirstName(), updated.getFirstName());
        assertEquals(update.getLastName(), updated.getLastName());
    }

    @Test
    public void testImmutablePermissions() {
        UserSpec builder = new UserSpec();
        builder.setPermissionIds(new Integer[]{
                userService.getPermission("group::administrator").getId()});

        Permission sup = userService.getPermission("group::administrator");

        User admin = userService.get("admin");
        userService.setPermissions(admin, Lists.newArrayList(sup));

        for (Permission p: userService.getPermissions(admin)) {
            logger.info("{}", p);
            if (p.getName().equals("admin") && p.getType().equals("user")) {
                // We found the user permission, all good
                return;
            }
        }
        throw new RuntimeException("The admin user was missing the user::admin permission");
    }

}

File: archivist/src/test/kotlin/com/zorroa/archivist/service/UserServiceTests.kt
package com.zorroa.archivist.service

import com.google.common.collect.Lists
import com.zorroa.archivist.AbstractTest
import com.zorroa.archivist.domain.User
import com.zorroa.archivist.domain.UserProfileUpdate
import com.zorroa.archivist.domain.UserSpec
import com.zorroa.archivist.sdk.security.Groups
import com.zorroa.sdk.client.exception.DuplicateEntityException
import org.junit.Before
import org.junit.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class UserServiceTests : AbstractTest() {

    internal lateinit var testUser: User

    @Before
    fun init() {
        val builder = UserSpec("billybob",
                "123password!",
                "testing@zorroa.com",
                firstName="BillyBob",
                lastName = "Rodriquez")
        testUser = userService.create(builder)
    }

    @Test
    fun createUser() {
        val builder = UserSpec("test",
                "123password",
                "test@test.com",
                firstName="Bilbo",
                lastName = "Baggins")

        builder.permissionIds =
                listOf(permissionService.getPermission(Groups.MANAGER).id)

        val user = userService.create(builder)

        /*
         * Re-authenticate this test as the user we just made.  Otherwise we can't
         * access the user folder.
         */
        authenticate(user.username)

        /*
         * Get the user's personal folder.
         */
        val folder = folderService.get("/Users/test")
        assertEquals(folder!!.name, user.username)

        /*
         * Get the permissions for the user.  This should contain the user's
         * personal permission and the user group
         */
        val perms = userService.getPermissions(user)
        assertEquals(3, perms.size.toLong())
        assertTrue(userService.hasPermission(user, permissionService.getPermission(Groups.MANAGER)))
        assertTrue(userService.hasPermission(user, permissionService.getPermission("user::test")))
        assertFalse(userService.hasPermission(user, permissionService.getPermission(Groups.DEV)))
    }

    @Test(expected = DuplicateEntityException::class)
    fun createDuplicateUser() {
        val builder = UserSpec("test",
                "123password",
                "test@test.com",
                firstName="Bilbo",
                lastName = "Baggins")
        userService.create(builder)
        userService.create(builder)
    }

    @Test
    fun testSetEnabled() {
        val builder = UserSpec("test",
                "123password",
                "test@test.com",
                firstName="Bilbo",
                lastName="Baggins")
        val user = userService.create(builder)
        assertTrue(userService.setEnabled(user, false))
    }

    @Test
    fun setPermissions() {
        val builder = UserSpec("bilbob",
                "123password",
                "test@test.com",
                firstName="Bilbo",
                lastName="Baggins")
        builder.permissionIds = listOf(permissionService.getPermission(Groups.MANAGER).id)

        val user = userService.create(builder)
        val dev = permissionService.getPermission(Groups.DEV)
        assertTrue(userService.hasPermission(user, permissionService.getPermission(Groups.MANAGER)))
        assertFalse(userService.hasPermission(user, dev))

        userService.setPermissions(user, Lists.newArrayList(dev))
        assertFalse(userService.hasPermission(user, permissionService.getPermission(Groups.MANAGER)))
        assertTrue(userService.hasPermission(user, dev))
    }

    @Test(expected = IllegalArgumentException::class)
    fun updatePasswordFailureTooShort() {
        try {
            userService.resetPassword(testUser, "nogood")
        } catch (e: IllegalArgumentException) {
            assertTrue(e.message!!.contains("or more characters"))
            throw e
        }

    }

    @Test(expected = IllegalArgumentException::class)
    fun updatePasswordFailureNoCaps() {
        try {
            userService.resetPassword(testUser, "nogood")
        } catch (e: IllegalArgumentException) {
            assertTrue(e.message!!.contains("upper case"))
            throw e
        }

    }

    @Test(expected = IllegalArgumentException::class)
    fun updatePasswordFailureNoNumbers() {
        try {
            userService.resetPassword(testUser, "nogood")
        } catch (e: IllegalArgumentException) {
            assertTrue(e.message!!.contains("least 1 number."))
            throw e
        }

    }

    @Test
    fun updatePassword() {
        userService.resetPassword(testUser, "DogCatched1")
        userService.checkPassword(testUser.username, "DogCatched1")
    }

    @Test
    fun updateUser() {
        val builder = UserSpec("test",
                "123password",
                "test@test.com",
                firstName="Bilbo",
                lastName = "Baggins")
        val user = userService.create(builder)

        val update = UserProfileUpdate()
        update.username = "bilbo"
        update.firstName = "foo"
        update.lastName = "bar"
        update.email = "test@test.com"

        assertTrue(userService.update(user, update))
        val updated = userService.get(user.id)
        assertEquals(update.username, updated.username)
        assertEquals(update.email, updated.email)
        assertEquals(update.firstName, updated.firstName)
        assertEquals(update.lastName, updated.lastName)
    }

    @Test
    fun updateUsername() {
        val builder = UserSpec("bilbob",
                "123password",
                "test@test.com",
                firstName="Bilbo",
                lastName = "Baggins")
        val user = userService.create(builder)

        val update = UserProfileUpdate("gandalf",
                "gandalf@zorroa.com",
                "Bilbo","Baggins")

        assertTrue(folderService.exists("/Users/bilbob"))
        assertTrue(permissionService.permissionExists("user::bilbob"))
        assertTrue(userService.update(user, update))
        assertFalse(folderService.exists("/Users/bilbob"))
        assertFalse(permissionService.permissionExists("user::bilbob"))
        assertTrue(folderService.exists("/Users/gandalf"))
        assertTrue(permissionService.permissionExists("user::gandalf"))
    }

    @Test
    fun updateUsernameWithBlankUsername() {

        val update = UserProfileUpdate("",
                "gandalf@zorroa.com",
                "Bilbo","Baggins")

        assertTrue(userService.update(testUser, update))
        assertEquals(testUser.username, userService.get(testUser.id).username)
        assertTrue(folderService.exists("/Users/billybob"))
        assertTrue(permissionService.permissionExists("user::billybob"))
    }

    @Test
    fun testImmutablePermissions() {
        val builder = UserSpec("test",
                "123password",
                "test@test.com",
                firstName="Bilbo",
                lastName = "Baggins")
        builder.permissionIds = listOf(permissionService.getPermission(Groups.ADMIN).id)
        val sup = permissionService.getPermission(Groups.ADMIN)

        val admin = userService.get("admin")
        userService.setPermissions(admin, Lists.newArrayList(sup))

        for (p in userService.getPermissions(admin)) {
            if (p.name == "admin" && p.type == "user") {
                // We found the user permission, all good
                return
            }
        }
        throw RuntimeException("The admin user was missing the user::admin permission")
    }

}
